package/MDAnalysis/analysis/diffusionmap.py

```diff
@@ -0,0 +1,377 @@
+
+
+class DistanceMatrix(AnalysisBase):
+    """ Calculate the pairwise distance between each frame in a trajectory using
+        a given metric
+
+    Attributes
+    ----------
+    atoms : AtomGroup
+        Selected atoms in trajectory subject to dimension reduction
+    dist_matrix : array
+        Array of all possible ij metric distances between frames in trajectory.
+        This matrix is symmetric with zeros on the diagonal.
+
+    Methods
+    -------
+    save(filename)
+        Save the `dist_matrix` to a given filename
+    """
+    def __init__(self, u, select='all', metric=rms.rmsd, cutoff=1E0-5,
+                 weights=None, start=None, stop=None, step=None):
+        """
+        Parameters
+        ----------
+        u : trajectory `~MDAnalysis.core.AtomGroup.Universe`
+            The MD Trajectory for dimension reduction, remember that
+            computational cost of eigenvalue decomposition
+            scales at O(N^3) where N is the number of frames.
+            Cost can be reduced by increasing step interval or specifying a
+            start and stop.
+        select: str, optional
+            Any valid selection string for
+            :meth:`~MDAnalysis.core.AtomGroup.AtomGroup.select_atoms`
+            This selection of atoms is used to calculate the RMSD between
+            different frames. Water should be excluded.
+        metric : function, optional
+            Maps two numpy arrays to a float, is positive definite and
+            symmetric, Default: metric is set to rms.rmsd().
+        cutoff : float, optional
+            Specify a given cutoff for metric values to be considered equal,
+            Default: 1EO-5
+        weights : array, optional
+            Weights to be given to coordinates for metric calculation
+        start : int, optional
+            First frame of trajectory to analyse, Default: 0
+        stop : int, optional
+            Last frame of trajectory to analyse, Default: -1
+        step : int, optional
+            Step between frames to analyse, Default: 1
+        """
+
+        self._u = u
+        traj = self._u.trajectory
+        self.atoms = self._u.select_atoms(select)
+        self._metric = metric
+        self._cutoff = cutoff
+        self._weights = weights
+        # remember that this must be called before referencing self.nframes
+        self._setup_frames(traj, start, stop, step)
+
+    def _prepare(self):
+        logger.info('numframes {0}'.format(self.nframes))
+        self.dist_matrix = np.zeros((self.nframes, self.nframes))
+
+    def _single_frame(self):
+        traj_index = self._ts.frame
+        i_ref = self.atoms.positions - self.atoms.center_of_mass()
+        # diagonal entries need not be calculated due to metric(x,x) == 0 in
+        # theory, _ts not updated properly. Possible savings by setting a
+        # cutoff for significant decimal places to sparsify matrix
+        for j in range(self.stop-1, self._ts.frame-1, -self.step):
```

Comment:
Oh, duh, that should definitely work. Thanks.

On Fri, Jun 17, 2016 at 3:06 PM kain88-de <notifications@github.com> wrote:

> In package/MDAnalysis/analysis/diffusionmap.py
> <https://github.com/MDAnalysis/mdanalysis/pull/863#discussion_r67581964>:
>
> > +        self._cutoff = cutoff
> > +        self._weights = weights
> > +        # remember that this must be called before referencing self.nframes
> > +        self._setup_frames(traj, start, stop, step)
> > +
> > +    def _prepare(self):
> > +        logger.info('numframes {0}'.format(self.nframes))
> > +        self.dist_matrix = np.zeros((self.nframes, self.nframes))
> > +
> > +    def _single_frame(self):
> > +        traj_index = self._ts.frame
> > +        i_ref = self.atoms.positions - self.atoms.center_of_mass()
> > +        # diagonal entries need not be calculated due to metric(x,x) == 0 in
> > +        # theory, _ts not updated properly. Possible savings by setting a
> > +        # cutoff for significant decimal places to sparsify matrix
> > +        for j in range(self.stop-1, self._ts.frame-1, -self.step):
>
> @jdetle <https://github.com/jdetle> reset the frame after the inner loop
> again
>
> for i, ts in enumerate(trj):
>     for j, ts in enumerate(trj[i:]):
>         # ... do stuff
>     trj[i] # reset to outerloop position
>
> This should help the outer iterator continue at the right position.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/863/files/319232c9116a676edb95fa8aed7e67e7a57ec16a#r67581964>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKcARvvah0eMKKoeEP6hbnMag_a-orw8ks5qMxpOgaJpZM4IpYIS>
> .
>
-- 
Have a wonderful day,
John J. Detlefs